homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,347 @@
+        self,
+        version: str | None = None,
+        backup: bool | None = None,
+    ) -> None:
+        """Install an update.
+
+        Version can be specified to install a specific version. When `None`, the
+        latest version needs to be installed.
+
+        The backup parameter indicates a backup should be taken before
+        installing the update.
+        """
+        await self.hass.async_add_executor_job(self.install, version, backup)
+
+    def install(
+        self,
+        version: str | None = None,
+        backup: bool | None = None,
+    ) -> None:
+        """Install an update.
+
+        Version can be specified to install a specific version. When `None`, the
+        latest version needs to be installed.
+
+        The backup parameter indicates a backup should be taken before
+        installing the update.
+        """
+        raise NotImplementedError()
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if (current_version := self.current_version) is None or (
+            latest_version := self.latest_version
+        ) is None:
+            return None
+
+        if latest_version not in (current_version, self.__skipped_version):
+            return STATE_ON
+        return STATE_OFF
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any] | None:
+        """Return state attributes."""
+        if (release_summary := self.release_summary) is not None:
+            release_summary = release_summary[:255]
+
+        # If entity supports progress, return the in_progress value.
+        # Otherwise, we use the internal progress value.
+        if self.supported_features & SUPPORT_PROGRESS:
+            in_progress = self.in_progress
+        else:
+            in_progress = self.__in_progress
+
+        return {
+            ATTR_CURRENT_VERSION: self.current_version,
+            ATTR_IN_PROGRESS: in_progress,
+            ATTR_LATEST_VERSION: self.latest_version,
+            ATTR_RELEASE_SUMMARY: release_summary,
+            ATTR_RELEASE_URL: self.release_url,
+            ATTR_SKIPPED_VERSION: self.__skipped_version,
+            ATTR_TITLE: self.title,
+        }
+
+    @final
+    async def async_install_with_progress(
+        self,
+        version: str | None = None,
+        backup: bool | None = None,
+    ) -> None:
+        """Install update and handle progress if needed.
+
+        Handles setting the in_progress state in case the entity doesn't
+        support it natively.
+        """
+        if not (progress_support := self.supported_features & SUPPORT_PROGRESS):
+            self.__in_progress = True
+            self.async_write_ha_state()
+
+        await self.async_install(version, backup)
+
+        if not progress_support:
+            self.__in_progress = False
+
+        # Clear skipped version
+        self.__skipped_version = None
+
+        self.async_write_ha_state()
```

Comment:
Note/Review to self:

This needs to handle exceptions raised from `async_install`. If the call raises, for whatever reason, the progress still needs to be reset.

Reply:
Done in [`6331ba9` (#68248)](https://github.com/home-assistant/core/pull/68248/commits/6331ba98146f627c3e8a124691f26de83a351c5d)